File: P1_201902235/calculoIMC.py
import psycopg2
import re
import time
from datetime import datetime
import os

def base_datos():
    try:
        conn = psycopg2.connect(host="localhost", dbname="parcial1", user="postgres", password="2008000070*a", port=5432)
        return conn
    except Exception as e:
        print("Error en la conexión:", e)
        return None
    
def obtener_fecha_hora():
    now = datetime.now()
    fecha = now.strftime("%Y-%m-%d") 
    hora = now.strftime("%H:%M:%S")  
    return fecha, hora

def menu():
    print("\n")
    print("1. Ingresar datos de usuario")
    print("2. Ejecución de programa")
    print("3. Borrado de datos")
    print("4. Salir")

def ingresar_datos_cliente():
    while True:
        print("\nIngrese nombre de usuario:")
        while True:
            nombre = input("Nombre del cliente: ")
            if re.match(r"^[A-Za-z\s]+$", nombre): 
                break
            else:
                print("Error: El nombre solo debe contener letras y espacios. Intenta nuevamente.")
        return nombre


def borrar():
    try:
        conn = base_datos()
        if conn:
            cursor = conn.cursor()
            cursor.execute("DELETE FROM peso_corporal;")
            conn.commit()
            conn.close()
            print("Todos los datos han sido borrados de la base de datos.")
    except Exception as e:
        print(f"Error al borrar datos: {e}")

def calcular_imc(peso, altura):
    return peso / (altura ** 2)

def imc():
    nombre_usuario = ingresar_datos_cliente()
    while True:
        sexo = input("Ingresa tu sexo (M/F): ").strip().upper() 
        if sexo in ["M", "F"]:
            break  
        else:
            print("Sexo no válido. Por favor, ingresa 'M' o 'F'.")
    
    peso = float(input("Ingresa tu peso en kilogramos: "))
    altura = float(input("Ingresa tu altura en metros: "))
    imc_calculado = calcular_imc(peso, altura)

    print(f"Tu Índice de Masa Corporal (IMC) es: {imc_calculado:.2f}")

    fecha, hora = obtener_fecha_hora()

    conn = base_datos()
    if conn:
        cursor = conn.cursor()
        cursor.execute("""
            INSERT INTO peso_corporal (nombre, peso, altura, genero, imc, fecha, hora)
            VALUES (%s, %s, %s, %s, %s, %s, %s)
        """, (nombre_usuario, peso, altura, sexo, imc_calculado, fecha, hora))
        conn.commit()
        conn.close()


def principal():
    while True:
        menu()
        opcion = input("\nSeleccione una opción: ")

        if opcion == "1":
            imc()

        elif opcion == "2":
            print("Ejecución del programa.")
        
        elif opcion == "3":
            borrar()

        elif opcion == "4":
            print("Saliendo del sistema...")
            time.sleep(1)
            break

        else:
            print("Opción no válida. Intenta nuevamente.")

principal()

File: P1_201902235/calculoIMC.m
clc;
clear;

pkg load database;

% Conexión a la base de datos
conn = pq_connect(setdbopts('dbname', 'parcial1', 'host', 'localhost', 'port', '5432', 'user', 'postgres', 'password', '2008000070*a'));


% Función para calcular el IMC
function [peso, altura, genero, imc] = calcular_imc()
    while true
        try
            genero = input('Ingrese su género (M/F): ', 's');
            if ~any(strcmp(genero, {'M', 'F'}))
                error('Género no válido. Debe ser "M" o "F".');
            end

            peso = input('Ingrese su peso en kilogramos: ');
            if peso <= 0
                error('El peso debe ser un número positivo.');
            end

            altura = input('Ingrese su altura en metros: ');
            if altura <= 0
                error('La altura debe ser un número positivo.');
            end

            imc = peso / (altura^2);
            fprintf('Su IMC es %.2f\n', imc);
            return;

        catch err
            fprintf('Error: %s\n', err.message);
        end
    end
end

% Función para registrar los datos en la base de datos
function registrar_imc(conn, nombre, peso, altura, genero, imc, fecha, hora)
    sql = 'INSERT INTO peso_corporal (nombre, peso, altura, genero, imc, fecha, hora) VALUES ($1, $2, $3, $4, $5, $6::date, $7::time)';
    pq_exec_params(conn, sql, {nombre, peso, altura, genero, imc, fecha, hora});
end

% Función para guardar los datos en un archivo de texto
function guardar_en_txt(nombre, peso, altura, genero, imc, fecha, hora)
    fecha_formateada = datestr(fecha, 'yyyy-mm-dd');
    hora_formateada = datestr(hora, 'HH:MM:SS');

    archivo = fopen('imc_registros.txt', 'a');
    if archivo == -1
        disp('Error al abrir el archivo de texto.');
        return;
    end

    fprintf(archivo, 'Nombre: %s | Peso: %.2f kg | Altura: %.2f m | Género: %s | IMC: %.2f | Fecha: %s | Hora: %s\n', nombre, peso, altura, genero, imc, fecha_formateada, hora_formateada);
    fclose(archivo);
end

% Función para mostrar el historial de la base de datos
function historial_datos(conn)
    sql_query = 'SELECT nombre, peso, altura, genero, imc, fecha, hora FROM peso_corporal;';
    N = pq_exec_params(conn, sql_query);

    if !isempty(N)
        disp('Historial de registros:');
        for i = 1:size(N, 1)
            % Convertir valores numéricos a formato adecuado si es necesario
            peso = str2double(N{i, 2});
            altura = str2double(N{i, 3});
            imc = str2double(N{i, 5});

            fprintf('Nombre: %s, Peso: %.2f, Altura: %.2f, Género: %s, IMC: %.2f, Fecha: %s, Hora: %s\n', ...
                    N{i, 1}, peso, altura, N{i, 4}, imc, N{i, 6}, N{i, 7});
        end
    else
        disp('No se encontraron registros.');
    end
end

function borrar_datos(conn)
    pq_exec_params(conn, 'DELETE FROM peso_corporal;');
    disp('Se han borrado los registros.');
end

% Programa principal
while true
    disp('------ Menú ------');
    disp('1. Calcular IMC');
    disp('2. Mostrar historial');
    disp('3. Borrar historial');
    disp('4. Salir');
    opcion = input('Seleccione una opción: ');

    switch opcion
        case 1
            % Ingreso de nombre y cálculo del IMC
            nombre = input('Ingrese su nombre: ', 's');
            [peso, altura, genero, imc] = calcular_imc();

            % Obtener la fecha y hora actuales
            fecha = datestr(now, 'yyyy-mm-dd');  % Fecha en formato YYYY-MM-DD
            hora = datestr(now, 'HH:MM:SS');     % Hora en formato HH:MM:SS

            % Registrar en la base de datos
            registrar_imc(conn, nombre, peso, altura, genero, imc, fecha, hora);
            disp('Resultado registrado en la base de datos.');

            % Guardar en el archivo de texto
            guardar_en_txt(nombre, peso, altura, genero, imc, fecha, hora);
            disp('Resultado guardado en el archivo de texto.');

        case 2
            % Mostrar historial
            historial_datos(conn);

        case 3
            % Borrar historial
            borrar_datos(conn);

        case 4
            disp('Saliendo del programa...');
            break;

        otherwise
            disp('Opción no válida. Intente nuevamente.');
    endswitch
endwhile

% Cerrar conexión a la base de datos
if exist("conn", "var")
    pq_close(conn);
else
    disp("No hay conexión abierta.");
end


